# Text Message Analysis Project

This portfolio project was completed as part of the requirements for the Codecademy.com Data Scientist - Natural Language Processing carreer path course.

The data used in this project consists of the National University of Singapore SMS Corpus dataset of text messages in English and Mandarin from Chen & Kan (2013).

## Import dataset

After importing the numpy and pandas packages, I imported the *clean_nus_sms.csv* file into Jupyter Notebook as a dataframe **df** and inspected it using the .head() and .info() commands.

The dataframe contains six columns: two contain integer data and four contain string data.

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv("clean_nus_sms.csv")

In [3]:
df.head(10)

,Unnamed: 0,id,Message,length,country,Date
0,0,10120,Bugis oso near wat...,21,SG,2003/4
1,1,10121,"Go until jurong point, crazy.. Available only ...",111,SG,2003/4
2,2,10122,I dunno until when... Lets go learn pilates...,46,SG,2003/4
3,3,10123,Den only weekdays got special price... Haiz......,140,SG,2003/4
4,4,10124,Meet after lunch la...,22,SG,2003/4
5,5,10125,m walking in citylink now ü faster come down.....,65,SG,2003/4
6,6,10126,5 nights...We nt staying at port step liao...T...,51,SG,2003/4
7,7,10127,Hey pple...$700 or $900 for 5 nights...Excelle...,81,SG,2003/4
8,8,10128,Yun ah.the ubi one say if ü wan call by tomorr...,160,SG,2003/4
9,9,10129,Hey tmr maybe can meet you at yck,33,SG,2003/4


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48598 entries, 0 to 48597
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  48598 non-null  int64 
 1   id          48598 non-null  int64 
 2   Message     48595 non-null  object
 3   length      48598 non-null  object
 4   country     48598 non-null  object
 5   Date        48598 non-null  object
dtypes: int64(2), object(4)
memory usage: 2.2+ MB


## Text preprocessing

### Data cleaning

To begin, I want to check for any null values and remove those from the dataframe.

In [5]:
df.isna().sum()

Unnamed: 0    0
id            0
Message       3
length        0
country       0
Date          0
dtype: int64

The **Message** column contains three null values. Let's go ahead and remove them from the dataframe.

In [6]:
df = df.dropna()

In [7]:
df.isna().sum()

Unnamed: 0    0
id            0
Message       0
length        0
country       0
Date          0
dtype: int64

The **Date** column in the dataframe displays a string consisting of a year and month. I want to split this into two columns--**Year** and **Month**--and convert the values into integers.

In [8]:
df[["Year", "Month"]] = df["Date"].str.split("/", expand=True).astype(int)

In [9]:
df.head()

,Unnamed: 0,id,Message,length,country,Date,Year,Month
0,0,10120,Bugis oso near wat...,21,SG,2003/4,2003,4
1,1,10121,"Go until jurong point, crazy.. Available only ...",111,SG,2003/4,2003,4
2,2,10122,I dunno until when... Lets go learn pilates...,46,SG,2003/4,2003,4
3,3,10123,Den only weekdays got special price... Haiz......,140,SG,2003/4,2003,4
4,4,10124,Meet after lunch la...,22,SG,2003/4,2003,4


In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 48595 entries, 0 to 48597
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  48595 non-null  int64 
 1   id          48595 non-null  int64 
 2   Message     48595 non-null  object
 3   length      48595 non-null  object
 4   country     48595 non-null  object
 5   Date        48595 non-null  object
 6   Year        48595 non-null  int32 
 7   Month       48595 non-null  int32 
dtypes: int32(2), int64(2), object(4)
memory usage: 3.0+ MB


Now that the values of **Date** have been split and converted into integers, that column can be deleted from the dataframe.

In [11]:
df = df.drop(columns=["Date"])

In [12]:
df.head()

,Unnamed: 0,id,Message,length,country,Year,Month
0,0,10120,Bugis oso near wat...,21,SG,2003,4
1,1,10121,"Go until jurong point, crazy.. Available only ...",111,SG,2003,4
2,2,10122,I dunno until when... Lets go learn pilates...,46,SG,2003,4
3,3,10123,Den only weekdays got special price... Haiz......,140,SG,2003,4
4,4,10124,Meet after lunch la...,22,SG,2003,4


The column labelled **length** refers to the character length of each text message included in the column **Message**.

First, I'll go ahead and rename the column label as **Characters**. Then, I will convert the string values of the column into integers.

In [13]:
df = df.rename(columns={"length": "Characters"})

In [14]:
df["Characters"] = df["Characters"].astype(int)

ValueError: invalid literal for int() with base 10: 'Err:511'

When I call *.astype(int)* to convert the values of **Characters** from strings to integers, I get a ValueError message: "invalid literal for int() with base 10: 'Err:511'".

Let's filter for "Err" to see how many rows feature an error value.

In [15]:
filtered_df = df[df['Characters'].str.contains("Err")]
print(filtered_df)

       Unnamed: 0     id  Message Characters        country  Year  Month
9468         9776  19896  Err:511    Err:511             SG  2003      4
18485       20557  10441  Err:509    Err:509  United States  2011      3
31264       34350  24234  Err:508    Err:508      Singapore  2011      4
32286       35529  25413  Err:510    Err:510      Singapore  2011      4


There four rows featuring error messages for the **Message** and **Characters** column. Let's remove these rows from the final dataframe.

In [16]:
df = df.drop([9468, 18485, 31264, 32286])

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 48591 entries, 0 to 48597
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  48591 non-null  int64 
 1   id          48591 non-null  int64 
 2   Message     48591 non-null  object
 3   Characters  48591 non-null  object
 4   country     48591 non-null  object
 5   Year        48591 non-null  int32 
 6   Month       48591 non-null  int32 
dtypes: int32(2), int64(2), object(3)
memory usage: 2.6+ MB


Now that the "Err" values have been removed from the dataframe, I can try to convert the values in **Characters** to integers again.

In [18]:
df["Characters"] = df["Characters"].astype(int)

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 48591 entries, 0 to 48597
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  48591 non-null  int64 
 1   id          48591 non-null  int64 
 2   Message     48591 non-null  object
 3   Characters  48591 non-null  int32 
 4   country     48591 non-null  object
 5   Year        48591 non-null  int32 
 6   Month       48591 non-null  int32 
dtypes: int32(3), int64(2), object(2)
memory usage: 2.4+ MB


Finally, I want to clean up the column **country**. Let's start by fixing this column label as well.

In [20]:
df = df.rename(columns={'country': 'Country'})

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 48591 entries, 0 to 48597
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  48591 non-null  int64 
 1   id          48591 non-null  int64 
 2   Message     48591 non-null  object
 3   Characters  48591 non-null  int32 
 4   Country     48591 non-null  object
 5   Year        48591 non-null  int32 
 6   Month       48591 non-null  int32 
dtypes: int32(3), int64(2), object(2)
memory usage: 2.4+ MB


Now let's go ahead and fix the values of that column. If you look back at **filtered_df**, you can see the values *SG* and *Singapore*, both of which refer to the nation Singapore.

I want to edit the country values to ensure all messages from a single country are properly attributed to the same country.

Let's begin by looking at the unique values of the **Country** column:

In [22]:
df.Country.value_counts()

Country
Singapore              22011
SG                      9803
India                   6901
United States           3748
USA                     1931
Sri Lanka               1017
Malaysia                 766
Pakistan                 751
unknown                  602
Canada                   198
Bangladesh               126
China                    107
india                    105
INDIA                     79
Philippines               67
Indonesia                 48
Nepal                     39
srilanka                  30
United Kingdom            30
Hungary                   28
Serbia                    22
Kenya                     20
Ghana                     18
Italia                    10
Turkey                    10
Trinidad and Tobago       10
Lebanon                   10
Slovenia                  10
Nigeria                   10
New Zealand               10
Macedonia                 10
UK                        10
Morocco                    9
Romania                    9
Austra

Let's go through each value and rename them so they are unified.

In [23]:
df['Country'] = df['Country'].replace({"SG": "Singapore", "unknown": "Unknown", "india": "India", "INDIA": "India", "srilanka": "Sri Lanka",
                                       "Italia": "Italy", "UK": "United Kingdom", "jamaica": "Jamaica", "BARBADOS": "Barbados", "MY": "Malaysia"})

In [24]:
df.Country.value_counts()

Country
Singapore              31814
India                   7085
United States           3748
USA                     1931
Sri Lanka               1047
Malaysia                 767
Pakistan                 751
Unknown                  602
Canada                   198
Bangladesh               126
China                    107
Philippines               67
Indonesia                 48
United Kingdom            40
Nepal                     39
Hungary                   28
Serbia                    22
Kenya                     20
Ghana                     18
Slovenia                  10
New Zealand               10
Nigeria                   10
Macedonia                 10
Lebanon                   10
Trinidad and Tobago       10
Italy                     10
Turkey                    10
Romania                    9
Morocco                    9
Australia                  9
Jamaica                    8
Barbados                   8
Spain                      5
France                     5
Name: 

In [ ]:
df.Message.